# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Statistics.cpp
# Commit: 7d099d4bac74
# Full Hash: 7d099d4bac7446d5bd798f813a2e78b1aefb1816
# Author: Denis Palmeiro <dpalmeiro@mozilla.com>
# Date: 2025-01-15 21:57:20
# Regressor Bug: 1932686
# File Overlap Count: 4
# Description:
#   Bug 1932686: Migrate enumeration legacy GC metrics to glean using a labeled counter. r=jonco,chutten
#   
#   Depends on D231645
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D231646
# ==============================================================================

diff -r 49990191195a -r 7d099d4bac74 js/src/gc/Statistics.cpp
--- a/js/src/gc/Statistics.cpp	Wed Jan 15 15:30:16 2025 +0000
+++ b/js/src/gc/Statistics.cpp	Wed Jan 15 15:30:16 2025 +0000
@@ -86,6 +86,11 @@
   }
 }
 
+JS_PUBLIC_API const char* JS::ExplainGCAbortReason(uint32_t val) {
+  GCAbortReason reason = static_cast<GCAbortReason>(val);
+  return js::gcstats::ExplainAbortReason(reason);
+}
+
 JS_PUBLIC_API bool JS::InternalGCReason(JS::GCReason reason) {
   return reason < JS::GCReason::FIRST_FIREFOX_REASON;
 }
@@ -176,6 +181,12 @@
   operator Phase() const { return phase; }
 };
 
+JS_PUBLIC_API const char* JS::GetGCPhaseName(uint32_t val) {
+  PhaseKind kind = static_cast<PhaseKind>(val);
+  MOZ_RELEASE_ASSERT(kind < PhaseKind::LIMIT);
+  return phaseKinds[kind].name;
+}
+
 static double t(TimeDuration duration) { return duration.ToMilliseconds(); }
 
 static TimeDuration TimeBetween(TimeStamp start, TimeStamp end) {
@@ -1343,18 +1354,21 @@
       // Record the longest phase in any long slice.
       if (wasLongSlice) {
         PhaseKind longest = LongestPhaseSelfTimeInMajorGC(slice.phaseTimes);
-        reportLongestPhaseInMajorGC(longest, [runtime](auto sample) {
-          runtime->metrics().GC_SLOW_PHASE(sample);
-        });
+        if (longest != PhaseKind::NONE) {
+          runtime->metrics().GC_SLOW_PHASE(phaseKinds[longest].telemetryBucket);
+          runtime->metrics().GC_GLEAN_SLOW_PHASE(
+              static_cast<uint32_t>(longest));
 
-        // If the longest phase was waiting for parallel tasks then record the
-        // longest task.
-        if (longest == PhaseKind::JOIN_PARALLEL_TASKS) {
-          PhaseKind longestParallel =
-              FindLongestPhaseKind(slice.maxParallelTimes);
-          reportLongestPhaseInMajorGC(longestParallel, [runtime](auto sample) {
-            runtime->metrics().GC_SLOW_TASK(sample);
-          });
+          // If the longest phase was waiting for parallel tasks then record the
+          // longest task.
+          if (longest == PhaseKind::JOIN_PARALLEL_TASKS) {
+            PhaseKind longestParallel =
+                FindLongestPhaseKind(slice.maxParallelTimes);
+            runtime->metrics().GC_SLOW_TASK(
+                phaseKinds[longestParallel].telemetryBucket);
+            runtime->metrics().GC_GLEAN_SLOW_TASK(
+                static_cast<uint32_t>(longestParallel));
+          }
         }
       }
     }
@@ -1364,14 +1378,6 @@
   }
 }
 
-template <typename Fn>
-void Statistics::reportLongestPhaseInMajorGC(PhaseKind longest, Fn reportFn) {
-  if (longest != PhaseKind::NONE) {
-    uint8_t bucket = phaseKinds[longest].telemetryBucket;
-    reportFn(bucket);
-  }
-}
-
 bool Statistics::startTimingMutator() {
   if (phaseStack.length() != 0) {
     // Should only be called from outside of GC.